### PR TITLE
Implementation of RFC 86: Column-oriented read API for vector layers

### PR DESCRIPTION
**Pull request by rouault**
Implements https://github.com/OSGeo/gdal/pull/5830.

Impacts:
- OGR core
- Python bindings
- Parquet driver
- GPKG driver
- FlatGeoBuf driver

Doc and tests updated

### CODE REVIEW

File: doc/source/tutorials/vector_api_tut.rst [hunk 1]
```diff
@@ -668,6 +668,293 @@ In Python:
...
+
+Reading From OGR using the Arrow C Stream data interface
+--------------------------------------------------------
+
+.. versionadded:: 3.6
+
+Instead of retrieving features one at a time, it is also possible to retrive
+them by batches, with a column-oriented memory layout, using the
+:cpp:func:`OGRLayer::GetArrowStream` method. Note that this method is more
+difficult to use than the traditionnal :cpp:func:`OGRLayer::GetNextFeature` approach,
```
Comment (nyalldawson) [inline doc/source/tutorials/vector_api_tut.rst:681]:
```suggestion
difficult to use than the traditional :cpp:func:`OGRLayer::GetNextFeature` approach,
```

File: doc/source/tutorials/vector_api_tut.rst [hunk 2]
```diff
@@ -668,6 +668,293 @@ In Python:
...
+no OGRLayer methods that affect the state of a layer should be called on a
+layer, while an ArrowArrayStream on it is active.
+
+The papszOptions that may be provided is a NULL terminated list of key=value
+strings, that may be driver specific.
+
+OGRLayer has a base implementation of GetArrowStream() that is such:
+
+- The get_schema() callback returns a schema whose top-level object returned is
+  of type Struct, and whose children consist in the FID column, all OGR attribute
```
Comment (nyalldawson) [inline doc/source/tutorials/vector_api_tut.rst:765]:
```suggestion
  of type Struct, and whose children consist of the FID column, all OGR attribute
```

**Comment by nyalldawson** [comment]
Aside from the above trivial typos, I've dug through this PR and can't spot any issues. (Admittedly there's a LOT of code here and it would be very easy to miss something :upside_down_face: )

File: doc/source/tutorials/vector_api_tut.rst (hunk 1 above)
Comment (rouault) [reply doc/source/tutorials/vector_api_tut.rst]: done

File: doc/source/tutorials/vector_api_tut.rst (hunk 2 above)
Comment (rouault) [reply doc/source/tutorials/vector_api_tut.rst]: done

**Comment by rouault** [comment]
@paleolimbot @jorisvandenbossche Want to have a chance testing that before I merge?

**Comment by rouault** [comment]
> (Doubling the metadata isn't strictly a problem and it might be Arrow C++ that's doing it

either Arrow C++ or the file itself, but that can be reproduced with pyarrow alone. For Parquet, if not forcing WKB encoding, OGR just delegates to libarrow/libparquet to return ArrowArray and ArrowSchema objects.
```
import pyarrow.parquet as pq
t = pq.read_table('nshn_water_line.parquet')
t.schema
[...]
geometry: list<linestrings: list<vertices: fixed_size_list<xyz: double>[3]>>
  child 0, linestrings: list<vertices: fixed_size_list<xyz: double>[3]>
      child 0, vertices: fixed_size_list<xyz: double>[3]
          child 0, xyz: double
      -- field metadata --
      ARROW:extension:name: 'geoarrow.point'
      ARROW:extension:metadata: 'crs�PROJCRS["NAD_1983_CSRS_20' + 924
      ARROW:extension:metadata: 'crs�PROJCRS["NAD_1983_CSRS_20' + 924
      ARROW:extension:name: 'geoarrow.point'
    -- field metadata --
    ARROW:extension:name: 'geoarrow.linestring'
    ARROW:extension:metadata: ''
    ARROW:extension:metadata: ''
    ARROW:extension:name: 'geoarrow.linestring'
  -- field metadata --
  ARROW:extension:metadata: ''
  ARROW:extension:name: 'geoarrow.multilinestring'
```


> I did get a crash when trying to read a bunch of large Parquet files sequentially

I tried looping over and over on the above file, and it's stable (RAM usage too)
```
from osgeo import ogr
ds = ogr.Open('nshn_water_line.parquet')
lyr = ds.GetLayer(0)
while True:
    stream = lyr.GetArrowStreamAsPyArrow()
    for record_batch in stream: pass
```

**Pull request merged**